Assets/GameEngine.cs

```diff
@@ -0,0 +1,43 @@
+﻿using UnityEngine;
+
+public class GameEngine : MonoBehaviour {
+
+    public GameObject unitPrefab;
+    private Transform source;
+
+	public void RegisterStrongholdClick(Transform selected) {
+        if (source == null) {
+            // If source is not yet selected, selected stronghold is the source
+            source = selected;
+        } else {
+            ObjectOwnerState sourceInfo = source.GetComponent<ObjectOwnerState>();
+
+            // If source is owned by a player, and not the same as target, send unit
+            if (sourceInfo.player != null && source != selected) {
+                SendUnit(sourceInfo, selected);
+            }
+
+            source = null;
+        }
+    }
+
+    private void SendUnit(ObjectOwnerState sourceInfo, Transform selected) {
+        // Create unit at source stronghold
+        GameObject unit = Instantiate(unitPrefab, source.position, Quaternion.identity);
+
+        int strongholdMagnitude = sourceInfo.magnitude;
+        int unitMagnitude = strongholdMagnitude / 2;
+
+        // Only send unit if it has a magnitude
+        if (unitMagnitude > 0) {
+            ObjectOwnerState unitInfo = unit.GetComponent<ObjectOwnerState>();
+
+            unitInfo.magnitude = unitMagnitude;
+            unitInfo.player = sourceInfo.player;
+            unit.GetComponent<UnitState>().SetSource(source.gameObject);
+            unit.GetComponent<Movement>().SetTarget(selected);
+
+            sourceInfo.magnitude -= unitMagnitude;
```

Comment:
Can you explain all of this code please

Reply:
Yep. So we instantiated the Unit object above, but did not assign it an owner or a magnitude. We do so here. We calculated `unitMagnitude` a few lines up _(half the stronghold magnitude)_, the owner is the owner of the source stronghold as that's where they're leaving from.

We also set the source and the target of the unit. The target is where it moves to _(there is a script that updates the position of the unit)_. The source is used in collision detection to ignore any collisions with the source while leaving the stronghold.

Finally, we subtract the unit magnitude _(as the unit is leaving)_ from the source stronghold's magnitude.